jax/interpreters/xla.py

```diff
@@ -1075,7 +1075,11 @@ def __iter__(self):
     if self.ndim == 0:
       raise TypeError("iteration over a 0-d array")  # same as numpy error
     else:
-      return self._value.__iter__()
+      device = self.device_buffer.device()
+      if device is None or device.platform == 'cpu':
```

Comment:
Under what conditions is device `None`?

Reply:
Good question! It's when the DeviceArray isn't backed by a device buffer at all but instead is a [lazy constant](https://github.com/google/jax/pull/1668), particularly [one of these](https://github.com/google/jax/blob/fddb28d8202dd1f1fc4cc78b44bccc1c82812746/jax/interpreters/xla.py#L1126-L1130), which also isn't sticky on a particular device (see #1916 for a description of what "sticky" means).

Luckily we shouldn't have to worry about that complexity much longer: after landing #3370 I hope to remove the lazy sublanguage completely.